Disk caching support

krishnasrinivas opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description
<!--- Describe your changes in detail -->
DiskCache for gateway and server

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
Gateway can use local disk for caching providing performance boost.

Note: Unit test cases are minimal as I don't know how we can unit test caching without manual intervention. I will see what other tests can be added.,If there are any suggestions on unit test cases please let me know.

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
manually

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

krishnasrinivas committed:
Disk caching support.

krishnasrinivas committed:
Refactor

Move Commit()/NoCommit() to diskCache
Remove diskCacheObject{} and use *os.File instead
Add diskCache.GetObjectInfo()
Anon access was not caching - fixed
Add --cache-dir command line support

krishnasrinivas committed:
Refactor

Add format.json support for cache
Command line args for --cache-dir --cache-max

krishnasrinivas committed:
Purge 'un-accessed" cached objects to free disk space

krishnasrinivas committed:
Enable disk caching for fs/xl

krishnasrinivas committed:
bug fixes

krishnasrinivas committed:
Hound-CI fixes. Add comments.

krishnasrinivas committed:
add --cache-dir --cache-max.
Refactor common code between GetObject and AnonGetObject

krishnasrinivas committed:
Refactor common code between GetObjectInfo and AnonGetObjectInfo

krishnasrinivas committed:
Enforce version checking

krishnasrinivas committed:
Support for caching during putobject

krishnasrinivas committed:
Use boltdb to store atime used for cache-eviction

krishnasrinivas committed:
Fix lint errors

harshavardhana commented on cmd/disk-cache.go:
```diff
@@ -9,14 +9,29 @@ import (
 	"encoding/json"
 )
 
+const diskCacheBackendVersion = "1"
```
Can you put this like and add a comment for each constant. 

```
const (
      diskCacheBackendVersion = "1"
      diskCacheObjectMetaVersion = "1.0.0"
)
```

harshavardhana commented on cmd/disk-cache.go:
```diff
@@ -9,14 +9,29 @@ import (
+const diskCacheObjectMetaVersion = "1.0.0"
+
+type diskCacheBoltdbEntry map[string]string
+
```
Add comments..

krishnasrinivas committed:
Use Fallocate for cache files
Delete entries from boltdb after purging files
Support caching for presigned-POST

krishnasrinivas committed:
Open DB TX in rw mode

krishnasrinivas committed:
Refactor and bugfixes

krishnasrinivas committed:
Rename MD5Sum to ETag

krishnasrinivas committed:
Refactor diskCache.Put

krishnasrinivas committed:
Cleanup

krishnasrinivas committed:
Add comments

krishnasrinivas committed:
Protect disk cache using locks

krishnasrinivas committed:
Cleanup

krishnasrinivas committed:
Add support for PutObject for server

krishnasrinivas committed:
If object is removed in the backend then remove from cache as well

krishnasrinivas committed:
fix lint errors

krishnasrinivas committed:
Cleanup

krishnasrinivas committed:
Cleanup and comments

krishnasrinivas committed:
Add BackendDown APIErrCode

krishnasrinivas committed:
govendor add boltdb

krishnasrinivas committed:
Add comments

harshavardhana commented on cmd/api-router.go:
```diff
@@ -21,13 +21,22 @@ import router "github.com/gorilla/mux"
+	var err error
+	if globalCacheDir != "" {
+		cache, err = newServerCacheObjects(globalObjectAPI, globalCacheDir, globalCacheMax, 0)
+		fatalIf(err, "Unable to init cache")
```
"Unable to initialize disk caching."

krishnasrinivas replied:
done

harshavardhana commented on cmd/disk-cache_test.go:
```diff
@@ -0,0 +1,88 @@
+package cmd
```
License header missing.

krishnasrinivas replied:
done

harshavardhana commented on cmd/gateway-main.go:
```diff
@@ -213,6 +217,9 @@ func gatewayMain(ctx *cli.Context) {
 	initNSLock(false) // Enable local namespace lock.
 
+	globalCacheDir = ctx.String("cache-dir")
+	globalCacheMax = ctx.Int("cache-max")
```
Mention what does this signify, both the options with a comment.

harshavardhana commented on cmd/gateway-router.go:
```diff
@@ -50,11 +50,16 @@ type gatewayAPIHandlers struct {
 // registerAPIRouter - registers S3 compatible APIs.
 func registerGatewayAPIRouter(mux *router.Router, gw GatewayLayer) {
+	cache, err := newGatewayCacheObjects(gw, globalCacheDir, globalCacheMax, 0)
+	if err != nil {
```
No need to do 'if err != nil`

harshavardhana commented on cmd/gateway-s3.go:
```diff
@@ -52,6 +54,14 @@ func s3ToObjectError(err error, params ...string) error {
 		object = params[1]
 	}
 
+	var urlErr *url.Error
```
This is more complete. 

```
// isNetworkError - is network error.. 
func isNetworkError(err error) bool {
        if err == nil {
                return false
        }
        switch err.(type) {
        case net.Error:
                switch err.(type) {
                case *net.DNSError, *net.OpError, net.UnknownNetworkError:
                        return true
                case *url.Error:
                        // For a URL error, where it replies back "connection closed"
                        if strings.Contains(err.Error(), "Connection closed by foreign host") {
                                return true
                        }
                default:
                        if strings.Contains(err.Error(), "net/http: TLS handshake timeout") {
                                // If error is - tlsHandshakeTimeoutError,.
                                return true
                        } else if strings.Contains(err.Error(), "i/o timeout") {
                                // If error is - tcp timeoutError.
                                return true
                        } else if strings.Contains(err.Error(), "connection timed out") {
                                // If err is a net.Dial timeout.
                                return true
                        }
                }
        }
        return false
}
```

harshavardhana commented on cmd/globals.go:
```diff
@@ -114,6 +114,9 @@ var (
 	globalActiveCred         credential
 	globalPublicCerts        []*x509.Certificate
 	globalXLObjCacheDisabled bool
+
```
Comments needed.. for each global option.

harshavardhana commented on cmd/server-main.go:
```diff
@@ -33,6 +33,16 @@ var serverFlags = []cli.Flag{
+	cli.StringFlag{
+		Name:  "cache-max",
+		Value: "80",
+		Usage: "Percent of disk space for Cache.",
```
"Percent of disk space used for caching."

harshavardhana commented on cmd/server-main.go:
```diff
@@ -33,6 +33,16 @@ var serverFlags = []cli.Flag{
+	cli.StringFlag{
+		Name:  "cache-dir",
+		Value: "",
+		Usage: "Cache directory.",
```
"Location of cache directory"

harshavardhana commented on cmd/server-main.go:
```diff
@@ -250,6 +263,13 @@ func serverMain(ctx *cli.Context) {
 	// Initialize name space lock.
 	initNSLock(globalIsDistXL)
 
+	newObject, err := newObjectLayer(globalEndpoints)
```
Moving this here won't work for distributed setup. For the disks to connect you need HTTP service to run first.

harshavardhana replied:
If no good reason then move back to after HTTP server has initialized.

krishnasrinivas replied:
moved back but changes were needed to newServerCacheObjects to accomodate the move back

harshavardhana reviewed (changes requested):
First round of review.. haven't looked into disk-cache.go yet.

harshavardhana commented on cmd/disk-cache.go:
```diff
@@ -0,0 +1,674 @@
+package cmd
```
License header missing.

harshavardhana commented on cmd/disk-cache.go:
```diff
@@ -0,0 +1,674 @@
+
+// Retusn if the disk usage is high.
+// Disk usage is high if disk used is > maxUsage
+func (c diskCache) diskUsageHigh() bool {
```
Can above two functions be combined? as single function which basically gives high/low behavior based on a boolean input?

krishnasrinivas replied:
no, its more readable with Low/High words than with boolean, also implementation is clearer with separate functions

harshavardhana commented on cmd/disk-cache.go:
```diff
@@ -0,0 +1,674 @@
+	bucket := objInfo.Bucket
+	object := objInfo.Name
+
+	objectLock := globalNSMutex.NewNSLock(diskCacheLockingPrefix+bucket, object)
```
why not pathJoin here? rather than +

harshavardhana commented on cmd/disk-cache.go:
```diff
@@ -0,0 +1,674 @@
+	defer objectLock.Unlock()
+
+	encPath := c.encodedPath(bucket, object)
+	if err := os.Rename(f.Name(), encPath); err != nil {
```
This won't work on windows.. renaming an active fd. You should use the `lock.LockedOpenFile()` which is allowed to be renamed in this manner.

harshavardhana commented on cmd/disk-cache.go:
```diff
@@ -0,0 +1,674 @@
+	if err != nil {
+		return nil, ObjectInfo{}, false, err
+	}
+	if err = json.Unmarshal(metaBytes, &objMeta); err != nil {
```
Use gson to unmarshal for better performance.. look at xl-v1-metadata.go and fs-v1-metadata.go

harshavardhana commented on cmd/disk-cache.go:
```diff
@@ -0,0 +1,674 @@
+		return nil, ObjectInfo{}, false, err
+	}
+	if objMeta.Version != diskCacheObjectMetaVersion {
+		return nil, ObjectInfo{}, false, errors.New("format not supported")
```
Use `errCorruptedFormat`

krishnasrinivas replied:
errCorruptedFormat will not be right because it gives the impression that the format file is itself corrupted. But this is a case where older binary running on newer formats.

harshavardhana replied:
Then use `errFormatNotSupported`

harshavardhana commented on cmd/disk-cache.go:
```diff
@@ -0,0 +1,674 @@
+// Inits the cache-dir if it is not init'ed already.
+// Initializing implies creation of cache-dir, data-dir, tmp-dir and format.json
+func newDiskCache(dir string, maxUsage, expiry int) (*diskCache, error) {
+	if err := os.MkdirAll(dir, 0766); err != nil {
```
Is it allowed to have simultaneous minio process to write to same cache location? if not can we prevent it by holding lock properly like how we do in `fs` ?

krishnasrinivas replied:
we do locking already in disk-cache

harshavardhana replied:
can you point me to the right location?

harshavardhana commented on cmd/disk-cache.go:
```diff
@@ -0,0 +1,674 @@
+	PutObjectFn         func(bucket, object string, size int64, data io.Reader, metadata map[string]string, sha256sum string) (objInfo ObjectInfo, err error)
+	AnonGetObjectFn     func(bucket, object string, startOffset int64, length int64, writer io.Writer) (err error)
+	AnonGetObjectInfoFn func(bucket, object string) (objInfo ObjectInfo, err error)
+	AnonPutObjectFn     func(bucket, object string, size int64, data io.Reader, metadata map[string]string, sha256sum string) (objInfo ObjectInfo, err error)
```
Is multipart not supported?

krishnasrinivas replied:
no

harshavardhana replied:
Can you write that why its not supported, so that someone doesn't implement it future.

harshavardhana commented on cmd/disk-cache.go:
```diff
@@ -0,0 +1,674 @@
+		GetObjectInfo = c.AnonGetObjectInfo
+	}
+	objInfo, err := GetObjectInfo(bucket, object)
+	_, backendDown := errorCause(err).(BackendDown)
```
This needs to be a function `isErrBackendDown`

krishnasrinivas replied:
using isNetworkOrHostDown() as function name, using implementation provided by you in one of the comments

harshavardhana commented on cmd/disk-cache.go:
```diff
@@ -0,0 +1,674 @@
+	_, backendDown := errorCause(err).(BackendDown)
+	if err != nil && !backendDown {
+		if _, ok := errorCause(err).(ObjectNotFound); ok {
+			// Delte the cached entry if backend object was deleted.
```
Typo `delete`

harshavardhana commented on cmd/disk-cache.go:
```diff
@@ -0,0 +1,674 @@
+			if anonReq {
+				if anon {
+					_, err = io.Copy(writer, io.NewSectionReader(r, startOffset, length))
+					return err
```
Highly nested if conditions, can you find a way to simplify this?

krishnasrinivas replied:
yes this piece is highly nested, could not simplify it further

harshavardhana reviewed (changes requested):
Initial review for disk-cache.go

vadmeste commented on cmd/gateway-handlers.go:
```diff
@@ -243,28 +255,44 @@ func (api gatewayAPIHandlers) PutObjectHandler(w http.ResponseWriter, r *http.Re
+		if api.cache != nil {
+			putObject = api.cache.AnonPutObject
+		}
+		// objInfo, err = objectAPI.AnonPutObject(bucket, object, size, r.Body, metadata, sha256sum)
```
Useless comment ?

krishnasrinivas committed:
Review fixes

harshavardhana commented on cmd/disk-cache_test.go:
```diff
@@ -1,3 +1,19 @@
+/*
+ * Minio Cloud Storage, (C) 2016 Minio, Inc.
```
s/2016/2017

krishnasrinivas committed:
Review fixes

harshavardhana commented on cmd/disk-cache.go:
```diff
@@ -0,0 +1,680 @@
+		return err
+	}
+	r, cachedObjInfo, anon, err := c.dcache.Get(bucket, object)
+	if err == nil {
```
In this condition is it okay to ignore `err != nil` ? and we should definitely simplify the following code.

harshavardhana commented on cmd/disk-cache.go:
```diff
@@ -0,0 +1,680 @@
+		}
+		cachedObjInfo, anon, infoErr := c.dcache.GetObjectInfo(bucket, object)
+		if infoErr == nil {
+			if !anonReq {
```
What is the difference between anon and anonReq?  perhaps we can choose a better variable name to differentiate their purpose.

harshavardhana reviewed (changes requested):
Larger change here is boltdb is not maintained anymore https://github.com/boltdb/bolt#a-message-from-the-author 

We should move to the fork by coreos - https://github.com/coreos/bbolt
